src/gasman.c

```diff
@@ -2081,54 +2081,75 @@ void CheckMasterPointers( void )
 }
 
 
-/****************************************************************************
-**
-*F  SwapMasterPoint( <bag1>, <bag2> ) . . . swap pointer of <bag1> and <bag2>
-*/
-void SwapMasterPoint (
-    Bag                 bag1,
-    Bag                 bag2 )
+// Swap the master pointers of bag1 and bag2
+//
+// We need to make sure the correct bags are garbage collected, so we always put
+// *both* bags on the ChangedBags linked-list, rather than pick through the
+// exact cases, as it is never incorrect to mark something changed.
+//
+// For completeness and future reference here are the necessary points to
+// consider.
+//
+// When swapping two master pointers, we have to take into account whether the
+// bags they refer to are marked as changed or not, as we otherwise may end up
+// in an inconsistent state, were a bag is referenced, but GASMAN does not know
+// this. GASMAN then collects this bag, resulting in a corrupted workspace.
+//
+// We consider the following three cases:
+//
+// 1. Both bags are old. Then if the original bag1 had been previously marked as
+//    changed (by having been put into the ChangedBags singly linked list), then
+//    we must make sure to mark the new bag2 as changed, too (and vice-versa).
+//
+// 2. Both bags are young. Then they typically will not be on the list of
+//    changed bags, as GAP implicitly treats all young bags as if they had been
+//    marked changed.
```

Comment:
Actually, I think now this sentence (which is of course mine, so all blame to me) is wrong or at least misleading. Perhaps just say: "... changed bags, as CHANGED_BAGS ignores young bags. However, young bags can still [... continue as before ]"

Reply:
In your defense, I rephrased some of the comments a bit, so this might be entirely my fault.